File: engine/services/webhook_service_test.go
package services

import (
	"testing"

	"kouma-engine/models"
	"kouma-engine/utils"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
)

func TestRenderTemplate(t *testing.T) {
	ctx := &WebhookPayloadContext{
		ProjectName:           "my-project",
		BuildVersion:          "v1.0",
		BuildResult:           "passed",
		BuildStatus:           "completed",
		BID:                   "bid-123",
		PID:                   "pid-456",
		CaseCount:             10,
		CasePassedCount:       8,
		CaseFailedCount:       1,
		CaseUndeterminedCount: 1,
		Timestamp:             "2024-01-01T00:00:00Z",
		Metadata: map[string]string{
			"branch": "main",
			"commit": "abc1234",
		},
	}

	tests := []struct {
		name     string
		template string
		expected string
	}{
		{
			name:     "simple replacement",
			template: `{"project":"{{projectName}}","status":"{{buildResult}}"}`,
			expected: `{"project":"my-project","status":"passed"}`,
		},
		{
			name:     "all variables",
			template: `{{projectName}} {{buildVersion}} {{buildResult}} {{buildStatus}} {{bid}} {{pid}} {{caseCount}} {{casePassedCount}} {{caseFailedCount}} {{caseUndeterminedCount}} {{timestamp}}`,
			expected: `my-project v1.0 passed completed bid-123 pid-456 10 8 1 1 2024-01-01T00:00:00Z`,
		},
		{
			name:     "no variables",
			template: `static text`,
			expected: `static text`,
		},
		{
			name:     "unknown variable",
			template: `{{unknown}}`,
			expected: `{{unknown}}`,
		},
		{
			name:     "metadata variable",
			template: `branch: {{meta.branch}}, commit: {{meta.commit}}`,
			expected: `branch: main, commit: abc1234`,
		},
		{
			name:     "metadata in JSON payload",
			template: `{"text":"Build on {{meta.branch}} ({{meta.commit}}): {{buildResult}}"}`,
			expected: `{"text":"Build on main (abc1234): passed"}`,
		},
		{
			name:     "unknown metadata key",
			template: `{{meta.nonexistent}}`,
			expected: `{{meta.nonexistent}}`,
		},
		{
			name:     "mixed standard and metadata variables",
			template: `{{projectName}} @ {{meta.branch}} -> {{buildResult}}`,
			expected: `my-project @ main -> passed`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := renderTemplate(tt.template, ctx)
			if result != tt.expected {
				t.Errorf("expected %q, got %q", tt.expected, result)
			}
		})
	}
}

func TestRenderTemplateNilMetadata(t *testing.T) {
	ctx := &WebhookPayloadContext{
		ProjectName: "my-project",
		BuildResult: "passed",
	}

	result := renderTemplate("{{projectName}} {{meta.branch}}", ctx)
	expected := "my-project {{meta.branch}}"
	if result != expected {
		t.Errorf("expected %q, got %q", expected, result)
	}
}

func TestParseJSONToParams(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected map[string]string
	}{
		{
			name:     "simple json",
			input:    `{"status":"passed","build":"bid-123"}`,
			expected: map[string]string{"status": "passed", "build": "bid-123"},
		},
		{
			name:     "empty json",
			input:    `{}`,
			expected: map[string]string{},
		},
		{
			name:     "not json",
			input:    `not json`,
			expected: map[string]string{},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := parseJSONToParams(tt.input)
			for k, v := range tt.expected {
				if result.Get(k) != v {
					t.Errorf("expected %s=%s, got %s", k, v, result.Get(k))
				}
			}
		})
	}
}

func TestSplitJSONPairs(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected int
	}{
		{
			name:     "two pairs",
			input:    `"key1":"val1","key2":"val2"`,
			expected: 2,
		},
		{
			name:     "comma in value",
			input:    `"key":"val,ue","key2":"val2"`,
			expected: 2,
		},
		{
			name:     "single pair",
			input:    `"key":"value"`,
			expected: 1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := splitJSONPairs(tt.input)
			if len(result) != tt.expected {
				t.Errorf("expected %d pairs, got %d: %v", tt.expected, len(result), result)
			}
		})
	}
}

func TestGetWebhooksByPID(t *testing.T) {
	mt := mtest.New(t, mtest.NewOptions().ClientType(mtest.Mock))

	mt.Run("success with webhooks", func(mt *mtest.T) {
		utils.MongoDB = mt.Client.Database("micoo")

		first := mtest.CreateCursorResponse(1, "micoo.webhooks", mtest.FirstBatch, bson.D{
			{"wid", "wid-1"},
			{"pid", "pid-123"},
			{"name", "Slack"},
			{"url", "https://hooks.slack.com/test"},
			{"method", "POST"},
			{"contentType", "json"},
			{"condition", "always"},
			{"payloadTemplate", `{"text":"test"}`},
			{"enabled", true},
		})
		killCursors := mtest.CreateCursorResponse(0, "micoo.webhooks", mtest.NextBatch)
		mt.AddMockResponses(first, killCursors)

		webhooks, err := GetWebhooksByPID("pid-123")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(webhooks) != 1 {
			t.Fatalf("expected 1 webhook, got %d", len(webhooks))
		}
		if webhooks[0].WID != "wid-1" {
			t.Errorf("expected wid wid-1, got %s", webhooks[0].WID)
		}
		if webhooks[0].Name != "Slack" {
			t.Errorf("expected name Slack, got %s", webhooks[0].Name)
		}
	})

	mt.Run("success with no webhooks", func(mt *mtest.T) {
		utils.MongoDB = mt.Client.Database("micoo")

		mt.AddMockResponses(mtest.CreateCursorResponse(0, "micoo.webhooks", mtest.FirstBatch))

		webhooks, err := GetWebhooksByPID("pid-123")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(webhooks) != 0 {
			t.Errorf("expected 0 webhooks, got %d", len(webhooks))
		}
	})

	mt.Run("find error", func(mt *mtest.T) {
		utils.MongoDB = mt.Client.Database("micoo")

		mt.AddMockResponses(
			mtest.CreateCommandErrorResponse(mtest.CommandError{
				Code:    123,
				Message: "find error",
			}),
		)

		_, err := GetWebhooksByPID("pid-123")
		if err == nil {
			t.Fatal("expected error, got nil")
		}
	})
}

func TestSendWebhooksForBuild_ConditionFiltering(t *testing.T) {
	build := &models.Build{
		PID:         "pid-123",
		BID:         "bid-456",
		BuildResult: "passed",
		BuildStatus: "completed",
	}
	project := &models.Project{
		PID:         "pid-123",
		ProjectName: "test-project",
	}

	ctx := &WebhookPayloadContext{
		ProjectName: project.ProjectName,
		BuildResult: build.BuildResult,
		BuildStatus: build.BuildStatus,
		BID:         build.BID,
		PID:         build.PID,
	}

	tests := []struct {
		condition  string
		result     string
		shouldSend bool
	}{
		{"always", "passed", true},
		{"always", "failed", true},
		{"always", "undetermined", true},
		{"success", "passed", true},
		{"success", "failed", false},
		{"success", "undetermined", false},
		{"fail", "failed", true},
		{"fail", "undetermined", true},
		{"fail", "passed", false},
	}

	for _, tt := range tests {
		t.Run(tt.condition+"_"+tt.result, func(t *testing.T) {
			ctx.BuildResult = tt.result
			shouldSend := tt.condition == "always" ||
				(tt.condition == "success" && ctx.BuildResult == "passed") ||
				(tt.condition == "fail" && (ctx.BuildResult == "failed" || ctx.BuildResult == "undetermined"))

			if shouldSend != tt.shouldSend {
				t.Errorf("condition=%s result=%s: expected shouldSend=%v, got %v", tt.condition, tt.result, tt.shouldSend, shouldSend)
			}
		})
	}
}

File: dashboard/backend/services/webhook-service.test.ts
import { describe, it, expect, mock, beforeEach } from 'bun:test';

mock.module('../models/webhook.ts', () => ({
    Webhook: Object.assign(
        function (this: Record<string, unknown>, data: Record<string, unknown>) {
            Object.assign(this, data);
            this.save = mock(() => Promise.resolve(this));
            return this;
        } as any,
        {
            find: mock(() => Promise.resolve([])),
            findOne: mock(() => Promise.resolve(null)),
            findOneAndUpdate: mock(() => Promise.resolve(null)),
            deleteOne: mock(() => Promise.resolve({ deletedCount: 1 })),
            deleteMany: mock(() => Promise.resolve({ deletedCount: 0 })),
        },
    ),
}));

import { Webhook } from '../models/webhook.ts';
import { webhookService } from './webhook-service.ts';

describe('WebhookService', () => {
    beforeEach(() => {
        (Webhook.find as any).mockClear();
        (Webhook.findOne as any).mockClear();
        (Webhook.findOneAndUpdate as any).mockClear();
        (Webhook.deleteOne as any).mockClear();
        (Webhook.deleteMany as any).mockClear();
    });

    describe('getWebhooksByPid', () => {
        it('should return webhooks for a project', async () => {
            const webhooks = [{ wid: 'w1', pid: 'p1', name: 'Slack' }];
            (Webhook.find as any).mockReturnValue({ lean: () => webhooks });

            const result = await webhookService.getWebhooksByPid('p1');

            expect(Webhook.find).toHaveBeenCalledWith({ pid: 'p1' });
            expect(result).toEqual(webhooks as any);
        });

        it('should return empty array when no webhooks', async () => {
            (Webhook.find as any).mockReturnValue({ lean: () => [] });

            const result = await webhookService.getWebhooksByPid('p1');
            expect(result).toEqual([]);
        });
    });

    describe('getWebhookByWid', () => {
        it('should return a webhook by wid', async () => {
            const webhook = { wid: 'w1', pid: 'p1', name: 'Slack' };
            (Webhook.findOne as any).mockResolvedValue(webhook);

            const result = await webhookService.getWebhookByWid('w1');

            expect(Webhook.findOne).toHaveBeenCalledWith({ wid: 'w1' });
            expect(result).toEqual(webhook as any);
        });

        it('should return null when not found', async () => {
            (Webhook.findOne as any).mockResolvedValue(null);

            const result = await webhookService.getWebhookByWid('nonexistent');
            expect(result).toBeNull();
        });
    });

    describe('deleteWebhook', () => {
        it('should delete a webhook by wid', async () => {
            (Webhook.deleteOne as any).mockResolvedValue({ deletedCount: 1 });

            const result = await webhookService.deleteWebhook('w1');

            expect(Webhook.deleteOne).toHaveBeenCalledWith({ wid: 'w1' });
            expect(result).toEqual({ deletedCount: 1 } as any);
        });
    });

    describe('deleteWebhooksByPid', () => {
        it('should delete all webhooks for a project', async () => {
            (Webhook.deleteMany as any).mockResolvedValue({ deletedCount: 3 });

            const result = await webhookService.deleteWebhooksByPid('p1');

            expect(Webhook.deleteMany).toHaveBeenCalledWith({ pid: 'p1' });
            expect(result).toEqual({ deletedCount: 3 } as any);
        });
    });

    describe('updateWebhook', () => {
        it('should update a webhook', async () => {
            const updated = { wid: 'w1', name: 'Updated' };
            (Webhook.findOneAndUpdate as any).mockResolvedValue(updated);

            const result = await webhookService.updateWebhook('w1', { name: 'Updated' });

            expect(Webhook.findOneAndUpdate).toHaveBeenCalledWith({ wid: 'w1' }, { $set: { name: 'Updated' } }, { new: true });
            expect(result).toEqual(updated as any);
        });
    });

    describe('renderTemplate', () => {
        const context = {
            projectName: 'my-project',
            buildVersion: 'v1.0.0',
            buildResult: 'passed',
            buildStatus: 'completed',
            bid: 'bid-123',
            pid: 'pid-456',
            caseCount: 10,
            casePassedCount: 8,
            caseFailedCount: 1,
            caseUndeterminedCount: 1,
            timestamp: '2024-01-01T00:00:00Z',
            metadata: { branch: 'main', commit: 'abc1234' },
        };

        it('should replace all known variables', () => {
            const template = '{{projectName}} {{buildVersion}} {{buildResult}}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('my-project v1.0.0 passed');
        });

        it('should handle numeric variables', () => {
            const template = '{{caseCount}} cases: {{casePassedCount}} passed, {{caseFailedCount}} failed';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('10 cases: 8 passed, 1 failed');
        });

        it('should replace unknown variables with empty string', () => {
            const template = '{{unknown}} text';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe(' text');
        });

        it('should handle template with no variables', () => {
            const result = webhookService.renderTemplate('static text', context);
            expect(result).toBe('static text');
        });

        it('should handle JSON payload template', () => {
            const template = '{"text":"Build {{buildVersion}} for {{projectName}}: {{buildResult}}"}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('{"text":"Build v1.0.0 for my-project: passed"}');
        });

        it('should replace metadata variables with {{meta.KEY}} syntax', () => {
            const template = 'branch: {{meta.branch}}, commit: {{meta.commit}}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('branch: main, commit: abc1234');
        });

        it('should handle metadata in JSON payload', () => {
            const template = '{"text":"Build on {{meta.branch}} ({{meta.commit}}): {{buildResult}}"}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('{"text":"Build on main (abc1234): passed"}');
        });

        it('should preserve unknown metadata key placeholder', () => {
            const template = '{{meta.nonexistent}}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('{{meta.nonexistent}}');
        });

        it('should preserve metadata placeholders when context has no metadata', () => {
            const contextNoMeta = { ...context, metadata: undefined };
            const template = '{{projectName}} {{meta.branch}}';
            const result = webhookService.renderTemplate(template, contextNoMeta);
            expect(result).toBe('my-project {{meta.branch}}');
        });

        it('should handle mixed standard and metadata variables', () => {
            const template = '{{projectName}} @ {{meta.branch}} -> {{buildResult}}';
            const result = webhookService.renderTemplate(template, context);
            expect(result).toBe('my-project @ main -> passed');
        });
    });
});
